File: frontend/src/types/finance.ts
/**
 * TypeScript types for Phase 3 Finance API responses.
 */

export interface FinancialAssumptions {
  emergency_reserve_pct: number   // 0.125 default
  working_capital_pct: number     // 0.20 default
  monthly_revenue_growth: number  // 0.02 default
  monthly_expense_growth: number  // 0.005 default
  fixed_cost_ratio: number        // 0.55 default
  variable_cost_ratio?: number | null
}

export interface AnalyzeRequest {
  business_id: string
  available_capital: number
  assumptions: FinancialAssumptions
}

export interface CashFlowRequest {
  business_id: string
  initial_monthly_revenue: number
  initial_monthly_expenses: number
  months: number
  monthly_revenue_growth: number
  monthly_expense_growth: number
  ramp_up_months: number
  ramp_up_factor: number
}

export interface InvestmentAllocationOut {
  equipment: number
  initial_inventory: number
  business_setup: number
  licensing: number
  marketing: number
  working_capital: number
  emergency_reserve: number
  total_allocated: number
  available_capital: number
  funding_gap: number
  is_feasible: boolean
  allocation_dict: Record<string, number>
}

export interface ScenarioOut {
  name: string
  monthly_revenue: number
  monthly_expenses: number
  monthly_profit: number
  annual_revenue: number
  annual_profit: number
  profit_margin_pct: number
}

export interface BreakEvenOut {
  fixed_costs_monthly: number
  variable_cost_ratio: number
  contribution_margin_ratio: number
  break_even_revenue: number
  assumed: boolean
}

export interface CashFlowMonthOut {
  month: number
  revenue: number
  expenses: number
  profit: number
  cumulative_cash_flow: number
}

export interface HealthScoreOut {
  budget_sufficiency: number
  profitability: number
  roi_score: number
  payback_score: number
  emergency_reserve_score: number
  expense_ratio_score: number
  total: number
  status: string
  strengths: string[]
  risks: string[]
  recommendations: string[]
}

export interface RiskIndicatorOut {
  name: string
  level: 'Low' | 'Medium' | 'High'
  explanation: string
}

export interface FullAnalysisOut {
  business_id: string
  business_name: string
  available_capital: number
  investment: InvestmentAllocationOut
  conservative: ScenarioOut
  expected: ScenarioOut
  optimistic: ScenarioOut
  roi_pct: number
  payback_months: number | null
  payback_feasible: boolean
  payback_note: string
  break_even: BreakEvenOut
  health: HealthScoreOut
  risks: RiskIndicatorOut[]
  cash_flow: CashFlowMonthOut[]
  disclaimer: string
}

export interface CashFlowOut {
  business_id: string
  months: CashFlowMonthOut[]
  disclaimer: string
}

File: backend/app/schemas/finance.py
"""
Pydantic schemas for Phase 3 Finance API.
"""
from typing import Dict, List, Optional
from pydantic import BaseModel, Field, field_validator


# ─────────────────────────────────────────────────────────────────────────────
# Request schemas
# ─────────────────────────────────────────────────────────────────────────────

class FinancialAssumptions(BaseModel):
    """Optional user-overridable assumptions for the financial engine."""
    emergency_reserve_pct:    float = Field(0.125, ge=0, le=0.5,  description="Fraction of capital held as reserve (default 12.5%)")
    working_capital_pct:      float = Field(0.20,  ge=0, le=0.5,  description="Fraction of capital as working capital (default 20%)")
    monthly_revenue_growth:   float = Field(0.02,  ge=0, le=0.30, description="Monthly revenue growth rate (default 2%)")
    monthly_expense_growth:   float = Field(0.005, ge=0, le=0.20, description="Monthly expense growth rate (default 0.5%)")
    fixed_cost_ratio:         float = Field(0.55,  ge=0.1, le=0.9, description="Fraction of expenses that are fixed (default 55%)")
    variable_cost_ratio: Optional[float] = Field(None, ge=0.01, le=0.99, description="Variable cost ratio (auto-derived if not set)")


class AnalyzeRequest(BaseModel):
    business_id:       str
    available_capital: float = Field(..., gt=0, description="User's available investment capital in INR")
    assumptions:       FinancialAssumptions = Field(default_factory=FinancialAssumptions)

    @field_validator("available_capital")
    @classmethod
    def capital_positive(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("available_capital must be positive")
        return v


class CashFlowRequest(BaseModel):
    business_id:              str
    initial_monthly_revenue:  float = Field(..., gt=0)
    initial_monthly_expenses: float = Field(..., gt=0)
    months:                   int   = Field(12, ge=1, le=60)
    monthly_revenue_growth:   float = Field(0.02, ge=0, le=0.30)
    monthly_expense_growth:   float = Field(0.005, ge=0, le=0.20)
    ramp_up_months:           int   = Field(2, ge=0, le=6)
    ramp_up_factor:           float = Field(0.70, ge=0.1, le=1.0)


# ─────────────────────────────────────────────────────────────────────────────
# Response schemas
# ─────────────────────────────────────────────────────────────────────────────

class InvestmentAllocationOut(BaseModel):
    equipment:         float
    initial_inventory: float
    business_setup:    float
    licensing:         float
    marketing:         float
    working_capital:   float
    emergency_reserve: float
    total_allocated:   float
    available_capital: float
    funding_gap:       float
    is_feasible:       bool
    allocation_dict:   Dict[str, float]


class ScenarioOut(BaseModel):
    name:               str
    monthly_revenue:    float
    monthly_expenses:   float
    monthly_profit:     float
    annual_revenue:     float
    annual_profit:      float
    profit_margin_pct:  float


class BreakEvenOut(BaseModel):
    fixed_costs_monthly:       float
    variable_cost_ratio:       float
    contribution_margin_ratio: float
    break_even_revenue:        float
    assumed:                   bool


class CashFlowMonthOut(BaseModel):
    month:                int
    revenue:              float
    expenses:             float
    profit:               float
    cumulative_cash_flow: float


class HealthScoreOut(BaseModel):
    budget_sufficiency:       float
    profitability:            float
    roi_score:                float
    payback_score:            float
    emergency_reserve_score:  float
    expense_ratio_score:      float
    total:                    float
    status:                   str
    strengths:                List[str]
    risks:                    List[str]
    recommendations:          List[str]


class RiskIndicatorOut(BaseModel):
    name:        str
    level:       str
    explanation: str


class FullAnalysisOut(BaseModel):
    business_id:      str
    business_name:    str
    available_capital: float
    investment:       InvestmentAllocationOut
    conservative:     ScenarioOut
    expected:         ScenarioOut
    optimistic:       ScenarioOut
    roi_pct:          float
    payback_months:   Optional[float]
    payback_feasible: bool
    payback_note:     str
    break_even:       BreakEvenOut
    health:           HealthScoreOut
    risks:            List[RiskIndicatorOut]
    cash_flow:        List[CashFlowMonthOut]
    disclaimer:       str


class CashFlowOut(BaseModel):
    business_id: str
    months:      List[CashFlowMonthOut]
    disclaimer:  str = "Financial projections are estimates for planning purposes only."
